feat(ingest): support view lineage for all sqlalchemy sources

mayurinehate opened the pull request:
Additional Changes:
1. Support incremental lineage for all sqlalchemy sources
2. Keep column level lineage enabled and incremental lineage disabled by default
3. Monkey-patch hive dialect to extract hive view definitions to extract lineage
4. Fix incremental_lineage_helper for empty upstreams
5. Support postgres-like partial view definitions

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

mayurinehate commented on metadata-ingestion/src/datahub/ingestion/source/sql/sql_common.py:
```diff
@@ -322,6 +335,18 @@ def __init__(self, config: SQLCommonConfig, ctx: PipelineContext, platform: str)
                 cached_domains=[k for k in self.config.domain], graph=self.ctx.graph
             )
 
+        self.views_processed: Set[str] = set()
```
Maintaining `views_processed` here , so that we have an option to emit lineage for remaining (failed) views as part of native source. For example postgres source already uses view dependency tables to get table level lineage. Once we support partial view definitions, we can follow logic - 
1. attempt table+column lineage here and update views_processed for successful extraction.
2. fallback to native lineage extraction for views that are not already processed (i.e. failed in lineage extraction).

hsheth2 replied:
What do you mean by "partial view definitions"?

mayurinehate replied:
If the view is created using SQL

`CREATE VIEW db1.array_struct_test_view as select * from db1.array_struct_test;`

Then the view definition as retrieved from database - and set as viewDefinition.viewLogic is like
```
select `array_struct_test`.`property_id`, `array_struct_test`.`service` from `db1`.`array_struct_test`
```

Observed this for postgres and presto-on-hive. This may also be the case for a few others.

hsheth2 replied:
Let's create a helper that leaves SqlParsingResult with query_type=CREATE as-is, but for type=SELECT it injects the out_table into the result

mayurinehate replied:
done

mayurinehate commented on metadata-ingestion/src/datahub/ingestion/source/sql/sql_common.py:
```diff
@@ -322,6 +335,18 @@ def __init__(self, config: SQLCommonConfig, ctx: PipelineContext, platform: str)
+            self._view_definition_cache = FileBackedDict[str]()
+        else:
+            self._view_definition_cache = {}
+
```
moved `use_file_backed_cache` related logic from teraform source here.

hsheth2 replied:
Now that we fixed the file backed dict to support windows, imo we can probably drop this config flag

mayurinehate replied:
Can you point me to the PR that fixes this ?

hsheth2 commented on metadata-ingestion/src/datahub/configuration/common.py:
```diff
@@ -283,7 +283,7 @@ class VersionedConfig(ConfigModel):
 class LineageConfig(ConfigModel):
     incremental_lineage: bool = Field(
-        default=True,
+        default=False,
```
For dbt, we need to make sure the default is still enabled

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/sql/hive.py:
```diff
@@ -90,19 +96,39 @@ def dbapi_get_columns_patched(self, connection, table_name, schema=None, **kw):
 
 
+try:
+    from pyhive.sqlalchemy_hive import HiveDialect
```
This seems fine for now, and we can fix up when we refactor sql common next week

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/sql/hive.py:
```diff
@@ -90,19 +96,39 @@ def dbapi_get_columns_patched(self, connection, table_name, schema=None, **kw):
+    HiveDialect.get_view_definition = get_view_definition_patched
+except ModuleNotFoundError:
+    pass
+except Exception as e:
```
Failure to patch should cause the source to fail to load right?

mayurinehate replied:
Right. Let me remove this exception handling.

mayurinehate replied:
done

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/sql/two_tier_sql_source.py:
```diff
@@ -71,6 +71,10 @@ def __init__(self, config, ctx, platform):
         super().__init__(config, ctx, platform)
         self.config: TwoTierSQLAlchemyConfig = config
 
+    def get_db_schema(self, dataset_identifier: str) -> Tuple[Optional[str], str]:
```
When we do the refactoring, let's put identifiers in a real data class instead of joining into a string in one place and then splitting in other places

hsheth2 commented on metadata-ingestion/src/datahub/utilities/sqlglot_lineage.py:
```diff
@@ -588,6 +588,9 @@ def _schema_aware_fuzzy_column_resolve(
         logger.debug("sqlglot failed to annotate types: %s", e)
+    except sqlglot.errors.ParseError as e:
+        # This is not a fatal error, so we can continue.
+        logger.debug("sqlglot failed to parse types: %s", e)
```
Let's merge this into the previous statement - `except` for both exception types

mayurinehate replied:
done

hsheth2 commented:
There's one remaining thing around handling partial view definitions

hsheth2 commented:
@mayurinehate Seeing a real issue in the smoke test - looks like sql_common needs a dependency on sqlparse

```
tests/test_stateful_ingestion.py:5: in <module>
    from datahub.ingestion.source.sql.mysql import MySQLConfig, MySQLSource
../metadata-ingestion/src/datahub/ingestion/source/sql/mysql.py:18: in <module>
    from datahub.ingestion.source.sql.sql_common import (
../metadata-ingestion/src/datahub/ingestion/source/sql/sql_common.py:34: in <module>
    from datahub.emitter.sql_parsing_builder import SqlParsingBuilder
../metadata-ingestion/src/datahub/emitter/sql_parsing_builder.py:11: in <module>
    from datahub.ingestion.source.usage.usage_common import BaseUsageConfig, UsageAggregator
../metadata-ingestion/src/datahub/ingestion/source/usage/usage_common.py:41: in <module>
    from datahub.utilities.sql_formatter import format_sql_query, trim_query
../metadata-ingestion/src/datahub/utilities/sql_formatter.py:4: in <module>
    import sqlparse
E   ModuleNotFoundError: No module named 'sqlparse'
```

mayurinehate commented:
> There's one remaining thing around handling partial view definitions

This is now added in this PR.



> @mayurinehate Seeing a real issue in the smoke test - looks like sql_common needs a dependency on sqlparse
 
Fixed it.

hsheth2 commented on metadata-ingestion/src/datahub/utilities/sqlglot_lineage.py:
```diff
@@ -1156,3 +1156,19 @@ def create_lineage_sql_parsed_result(
+    if result.query_type is QueryType.SELECT:
+        # Some platforms (e.g. postgres) store only <select statement> from view definition
+        # `create view V as <select statement>` . For such view definitions, `result.out_tables` and
+        # `result.column_lineage[].downstream` are empty in `sqlglot_lineage` response, whereas
```
looks like the comment got cut off here - not a huge thing, and we can fix in a follow up

mayurinehate replied:
updated.

hsheth2 commented:
The cypress test failure appears unrelated (`glossary/glossary_navigation.js`), so merging through
